File: pkg/ui/src/api/services/secrets-manager.ts
/**
 * Secrets Manager Service API Client
 * REST HTTP client for Secrets Manager via Go proxy
 * @module api/services/secrets-manager
 */

import { PROXY_BACKEND } from '@/config'
import { APIError } from '../client'

const api = PROXY_BACKEND.replace(/\/$/, '')

export class SecretsManagerService {
  async createSecret(params: {
    Name: string
    SecretString?: string
    SecretBinary?: string
    Description?: string
  }): Promise<any> {
    const res = await fetch(`${api}/secrets-manager/secrets`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(params),
    })
    if (!res.ok) throw new APIError(`Create secret failed`, res.status, 'secrets-manager')
    return res.json()
  }

  async getSecretValue(SecretId: string): Promise<any> {
    const res = await fetch(`${api}/secrets-manager/secrets/${encodeURIComponent(SecretId)}/value`)
    if (!res.ok) throw new APIError(`Get secret value failed`, res.status, 'secrets-manager')
    return res.json()
  }

  async putSecretValue(params: {
    SecretId: string
    SecretString?: string
    SecretBinary?: string
  }): Promise<any> {
    const { SecretId, ...body } = params
    const res = await fetch(`${api}/secrets-manager/secrets/${encodeURIComponent(SecretId)}/value`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    })
    if (!res.ok) throw new APIError(`Put secret value failed`, res.status, 'secrets-manager')
    return res.json()
  }

  async deleteSecret(SecretId: string, options?: {
    RecoveryWindowInDays?: number
    ForceDeleteWithoutRecovery?: boolean
  }): Promise<any> {
    const res = await fetch(`${api}/secrets-manager/secrets/${encodeURIComponent(SecretId)}`, {
      method: 'DELETE',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(options || {}),
    })
    if (!res.ok) throw new APIError(`Delete secret failed`, res.status, 'secrets-manager')
    return res.json()
  }

  async updateSecret(params: {
    SecretId: string
    SecretString?: string
    Description?: string
  }): Promise<any> {
    const { SecretId, ...body } = params
    const res = await fetch(`${api}/secrets-manager/secrets/${encodeURIComponent(SecretId)}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    })
    if (!res.ok) throw new APIError(`Update secret failed`, res.status, 'secrets-manager')
    return res.json()
  }

  async describeSecret(SecretId: string): Promise<any> {
    const res = await fetch(`${api}/secrets-manager/secrets/${encodeURIComponent(SecretId)}`)
    if (!res.ok) throw new APIError(`Describe secret failed`, res.status, 'secrets-manager')
    return res.json()
  }

  async listSecrets(options?: { MaxResults?: number }): Promise<any> {
    const res = await fetch(`${api}/secrets-manager/secrets`)
    if (!res.ok) throw new APIError(`List secrets failed`, res.status, 'secrets-manager')
    return res.json()
  }

  async rotateSecret(SecretId: string): Promise<any> {
    const res = await fetch(`${api}/secrets-manager/secrets/${encodeURIComponent(SecretId)}/rotate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
    })
    if (!res.ok) throw new APIError(`Rotate secret failed`, res.status, 'secrets-manager')
    return res.json()
  }

  async restoreSecret(SecretId: string): Promise<any> {
    const res = await fetch(`${api}/secrets-manager/secrets/${encodeURIComponent(SecretId)}/restore`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
    })
    if (!res.ok) throw new APIError(`Restore secret failed`, res.status, 'secrets-manager')
    return res.json()
  }

  async getRandomPassword(options?: { PasswordLength?: number }): Promise<any> {
    const res = await fetch(`${api}/secrets-manager/random-password`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(options || {}),
    })
    if (!res.ok) throw new APIError(`Get random password failed`, res.status, 'secrets-manager')
    return res.json()
  }
}

export const secretsManagerService = new SecretsManagerService()

export const createSecret = (params: Parameters<SecretsManagerService['createSecret']>[0]) =>
  secretsManagerService.createSecret(params)
export const getSecretValue = (SecretId: string) => secretsManagerService.getSecretValue(SecretId)
export const putSecretValue = (params: Parameters<SecretsManagerService['putSecretValue']>[0]) =>
  secretsManagerService.putSecretValue(params)
export const deleteSecret = (SecretId: string, options?: Parameters<SecretsManagerService['deleteSecret']>[1]) =>
  secretsManagerService.deleteSecret(SecretId, options)
export const updateSecret = (params: Parameters<SecretsManagerService['updateSecret']>[0]) =>
  secretsManagerService.updateSecret(params)
export const describeSecret = (SecretId: string) => secretsManagerService.describeSecret(SecretId)
export const listSecrets = (options?: Parameters<SecretsManagerService['listSecrets']>[0]) =>
  secretsManagerService.listSecrets(options)
export const rotateSecret = (SecretId: string) => secretsManagerService.rotateSecret(SecretId)
export const restoreSecret = (SecretId: string) => secretsManagerService.restoreSecret(SecretId)
export const getRandomPassword = (options?: Parameters<SecretsManagerService['getRandomPassword']>[0]) =>
  secretsManagerService.getRandomPassword(options)

export default secretsManagerService

File: pkg/proxy/internal/adapters/http/secretsmanager.go
package httphandlers

import (
	"net/http"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/secretsmanager"
	"github.com/go-chi/chi/v5"
)

func (h *ProxyHandler) registerSecretsManagerRoutes(r chi.Router) {
	r.Route("/secrets-manager", func(r chi.Router) {
		r.Get("/secrets", h.listSecrets)
		r.Post("/secrets", h.createSecret)
		r.Get("/secrets/{secretId}", h.describeSecret)
		r.Put("/secrets/{secretId}", h.updateSecret)
		r.Delete("/secrets/{secretId}", h.deleteSecret)
		r.Post("/secrets/{secretId}/restore", h.restoreSecret)
		r.Post("/secrets/{secretId}/rotate", h.rotateSecret)
		r.Put("/secrets/{secretId}/value", h.putSecretValue)
		r.Get("/secrets/{secretId}/value", h.getSecretValue)
		r.Post("/random-password", h.getRandomPassword)
	})
}

func (h *ProxyHandler) listSecrets(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &secretsmanager.ListSecretsInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.SecretsManager().ListSecrets(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to list secrets", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) createSecret(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &secretsmanager.CreateSecretInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.SecretsManager().CreateSecret(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to create secret", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) getSecretValue(w http.ResponseWriter, r *http.Request) {
	input := &secretsmanager.GetSecretValueInput{
		SecretId: aws.String(urlParam(r, "secretId")),
	}
	result, err := h.Svc.SecretsManager().GetSecretValue(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to get secret value", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) putSecretValue(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &secretsmanager.PutSecretValueInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input.SecretId = aws.String(urlParam(r, "secretId"))
	result, err := h.Svc.SecretsManager().PutSecretValue(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to put secret value", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) deleteSecret(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	var body struct {
		RecoveryWindowInDays        int64 `json:"RecoveryWindowInDays"`
		ForceDeleteWithoutRecovery bool   `json:"ForceDeleteWithoutRecovery"`
	}
	if err := parseBody(bodyBytes, &body); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input := &secretsmanager.DeleteSecretInput{
		SecretId: aws.String(urlParam(r, "secretId")),
	}
	if body.RecoveryWindowInDays > 0 {
		input.RecoveryWindowInDays = aws.Int64(body.RecoveryWindowInDays)
	}
	input.ForceDeleteWithoutRecovery = aws.Bool(body.ForceDeleteWithoutRecovery)
	result, err := h.Svc.SecretsManager().DeleteSecret(h.ctx, input)
	if err != nil {
		sendErrorWithStatus(w, "Failed to delete secret", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) describeSecret(w http.ResponseWriter, r *http.Request) {
	input := &secretsmanager.DescribeSecretInput{
		SecretId: aws.String(urlParam(r, "secretId")),
	}
	result, err := h.Svc.SecretsManager().DescribeSecret(h.ctx, input)
	if err != nil {
		sendErrorWithStatus(w, "Failed to describe secret", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) updateSecret(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &secretsmanager.UpdateSecretInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input.SecretId = aws.String(urlParam(r, "secretId"))
	result, err := h.Svc.SecretsManager().UpdateSecret(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to update secret", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) restoreSecret(w http.ResponseWriter, r *http.Request) {
	input := &secretsmanager.RestoreSecretInput{
		SecretId: aws.String(urlParam(r, "secretId")),
	}
	result, err := h.Svc.SecretsManager().RestoreSecret(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to restore secret", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) rotateSecret(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &secretsmanager.RotateSecretInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	input.SecretId = aws.String(urlParam(r, "secretId"))
	result, err := h.Svc.SecretsManager().RotateSecret(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to rotate secret", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}

func (h *ProxyHandler) getRandomPassword(w http.ResponseWriter, r *http.Request) {
	bodyBytes := readBody(r)
	input := &secretsmanager.GetRandomPasswordInput{}
	if err := parseBody(bodyBytes, input); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.Svc.SecretsManager().GetRandomPassword(h.ctx, input)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to get random password", err)
		return
	}
	writeJSON(w, http.StatusOK, result)
}
